src/pika_stream_base.cc

```diff
@@ -0,0 +1,616 @@
+      }
+    }
+
+    trim_ret.count += static_cast<int32_t>(filed_values.size());
+
+    // do the delete in batch
+    std::vector<std::string> fields_to_del;
+    fields_to_del.reserve(filed_values.size());
+    for (auto &fv : filed_values) {
+      fields_to_del.emplace_back(std::move(fv.field));
+    }
+    int32_t ret;
+    s = StreamStorage::DeleteStreamMessage(key, fields_to_del, ret, slot);
+    if (!s.ok()) {
+      return s;
+    }
+    assert(ret == fields_to_del.size());
+  }
+
+  s = storage::Status::OK();
+  return s;
+}
+
+storage::Status StreamStorage::TrimStream(int32_t &count, StreamMetaValue &stream_meta, const std::string &key,
+                                          StreamAddTrimArgs &args, const Slot *slot) {
+  assert(slot);
+  count = 0;
+  // 1 do the trim
+  TrimRet trim_ret;
+  storage::Status s;
+  if (args.trim_strategy == StreamTrimStrategy::TRIM_STRATEGY_MAXLEN) {
+    s = TrimByMaxlen(trim_ret, stream_meta, key, slot, args);
+  } else {
+    assert(args.trim_strategy == StreamTrimStrategy::TRIM_STRATEGY_MINID);
+    s = TrimByMinid(trim_ret, stream_meta, key, slot, args);
+  }
+
+  if (!s.ok()) {
+    return s;
+  }
+
+  if (trim_ret.count == 0) {
+    return s;
+  }
+
+  // 2 update stream meta
+  streamID first_id;
+  streamID max_deleted_id;
+  if (stream_meta.length() == trim_ret.count) {
+    // all the message in the stream were deleted
+    first_id = kSTREAMID_MIN;
+  } else {
+    first_id.DeserializeFrom(trim_ret.next_field);
+  }
+  assert(!trim_ret.max_deleted_field.empty());
+  max_deleted_id.DeserializeFrom(trim_ret.max_deleted_field);
+
+  stream_meta.set_first_id(first_id);
+  if (max_deleted_id > stream_meta.max_deleted_entry_id()) {
+    stream_meta.set_max_deleted_entry_id(max_deleted_id);
+  }
+  stream_meta.set_length(stream_meta.length() - trim_ret.count);
+
+  count = trim_ret.count;
+
+  s = storage::Status::OK();
+  return s;
+}
+
+storage::Status StreamStorage::DestoryStreams(std::vector<std::string> &keys, const Slot *slot) {
+  assert(slot);
+  storage::Status s;
+  for (const auto &key : keys) {
+    // 1.try to get the stream meta
+    StreamMetaValue stream_meta;
+    s = StreamStorage::GetStreamMeta(stream_meta, key, slot);
+    if (s.IsNotFound()) {
+      continue;
+    } else if (!s.ok()) {
+      LOG(ERROR) << "Get stream meta failed";
+      return s;
+    }
+
+    // 2 destroy all the cgroup
+    // 2.1 find all the cgroups' meta
+    auto cgroup_tid = stream_meta.groups_id();
+    if (cgroup_tid != kINVALID_TREE_ID) {
+      // Korpse TODO: delete the cgroup, now we don't have groups.
+    }
+
+    // 3 delete stream meta
+    s = StreamStorage::DeleteStreamMeta(key, slot);
+    if (!s.ok()) {
+      return s;
+    }
+
+    // 4 delete stream data
+    s = StreamStorage::DeleteStreamData(key, slot);
+    if (!s.ok()) {
+      return s;
+    }
+  }
+
+  s = storage::Status::OK();
+  return s;
+}
+
+bool StreamUtils::string2uint64(const char *s, uint64_t &value) {
```

Comment:
这些string2int 和 int2string 有通用的函数,还是用通用的好点
https://github.com/OpenAtomFoundation/pika/blob/3611c71fee2496fcd5c989ed350dbab72a291426/src/pstd/include/pstd_string.h#L47

Reply:
我试着用了一下，其中这个转换 usined long 的函数好像有问题。
```
/* Convert a string into a unsigned long. Returns 1 if the string could be parsed into a
 * (non-overflowing) unsigned long, 0 otherwise. The value will be set to the parsed
 * value when appropriate. */
int string2int(const char* s, size_t slen, unsigned long* lval) {
  long long llval;

  if (string2int(s, slen, &llval) == 0) {
    return 0;
  }

  if (llval > static_cast<long long>(ULONG_MAX)) {
    return 0;
  }

  *lval = static_cast<unsigned long>(llval);
  return 1;
}
```
在 ` if (llval > static_cast<long long>(ULONG_MAX))` 这一行，如果 long long 和 unsigned long 都是 64 位，那么 `static_cast<long long>(ULONG_MAX)` 的结果会是 -1，因为其补码是全 1，强转后就变成了 -1，导致正确字符串转换也会返回 0（比如"123"）。
我测试下来也确实有这个问题。

Reply:
能说一下这个怎么复现吗, 我在win10, MinGW64下 没有复现出来. 因为这段代码被我移植到 4.0分支了, 如果有问题, 两边都需要改一下,谢谢

Reply:
好像没有其它地方使用这个重载，所以pika没办法复现，可以直接在单独的代码里面跑一下。
我是 ubuntu18.04，gcc编译器。
刚才跑了这段代码，可以复现。
```cpp
#include <sys/time.h>
#include <unistd.h>
#include <cctype>
#include <cfloat>
#include <climits>
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <sstream>
#include <string>

#include <algorithm>

/* Convert a string into a long long. Returns 1 if the string could be parsed
 * into a (non-overflowing) long long, 0 otherwise. The value will be set to
 * the parsed value when appropriate. */
int string2int(const char* s, size_t slen, long long* value) {
  const char* p = s;
  size_t plen = 0;
  int negative = 0;
  unsigned long long v;

  if (plen == slen) {
    return 0;
  }

  /* Special case: first and only digit is 0. */
  if (slen == 1 && p[0] == '0') {
    if (value) {
      *value = 0;
    }
    return 1;
  }

  if (p[0] == '-') {
    negative = 1;
    p++;
    plen++;

    /* Abort on only a negative sign. */
    if (plen == slen) {
      return 0;
    }
  }

  while (plen < slen && p[0] == '0') {
    p++;
    plen++;
  }

  if (plen == slen) {
    if (value) {
      *value = 0;
    }
    return 1;
  }

  /* First digit should be 1-9, otherwise the string should just be 0. */
  if (p[0] >= '1' && p[0] <= '9') {
    v = p[0] - '0';
    p++;
    plen++;
  } else if (p[0] == '0' && slen == 1) {
    *value = 0;
    return 1;
  } else {
    return 0;
  }

  while (plen < slen && p[0] >= '0' && p[0] <= '9') {
    if (v > (ULLONG_MAX / 10)) { /* Overflow. */
      return 0;
    }
    v *= 10;

    if (v > (ULLONG_MAX - (p[0] - '0'))) { /* Overflow. */
      return 0;
    }
    v += p[0] - '0';

    p++;
    plen++;
  }

  /* Return if not all bytes were used. */
  if (plen < slen) {
    return 0;
  }

  if (negative != 0) {
    if (v > (static_cast<unsigned long long>(-(LLONG_MIN + 1)) + 1)) { /* Overflow. */
      return 0;
    }
    if (value) {
      *value = static_cast<long long>(-v);
    }
  } else {
    if (v > LLONG_MAX) { /* Overflow. */
      return 0;
    }
    if (value) {
      *value = static_cast<long long>(v);
    }
  }
  return 1;
}

/* Convert a string into a unsigned long. Returns 1 if the string could be parsed into a
 * (non-overflowing) unsigned long, 0 otherwise. The value will be set to the parsed
 * value when appropriate. */
int string2int(const char* s, size_t slen, unsigned long* lval) {
  long long llval;

  if (string2int(s, slen, &llval) == 0) {
    return 0;
  }

  if (llval > static_cast<long long>(ULONG_MAX)) {
    return 0;
  }

  *lval = static_cast<unsigned long>(llval);
  return 1;
}

int main(int argc, const char** argv) {
  const char s[] = "123";
  uint64_t v = 0;
  if (string2int(s, strlen(s), &v)) {
    printf("v: %lu\n", v);
  } else {
    printf("unexpected error, static_cast<long long>(ULONG_MAX) : %lld\n", static_cast<long long>(ULONG_MAX));
  }
  return 0;
}
```

Reply:
我提了个 issue 在这里：https://github.com/OpenAtomFoundation/pika/issues/1956

Reply:
那我代码还是先暂时用自己的转换函数？因为有很多string 转 uint64_t 的地方，另外也没有看到 string  转 int 的函数，string2int 的三个重载是转成 long long, long, unsigned long。

Reply:
先用你自己写的吧  这个在类型处理上确实有问题